src/catchup/CatchupWork.cpp

```diff
@@ -0,0 +1,498 @@
+uint32_t
+firstNotFinishedCheckpoint(uint32_t lastClosedLedger,
+                           HistoryManager const& historyManager)
+{
+    auto result = checkpointContainingLedger(lastClosedLedger, historyManager);
+    if (lastClosedLedger < result)
+    {
+        return result;
+    }
+    else
+    {
+        return result + historyManager.getCheckpointFrequency();
+    }
+}
+
+// return first ledger that should be applied so at least count history entries
+// are stored in database (count 0 is changed to 1, because even doing only
+// bucket apply gives one entry)
+//
+// if it is impossible, returns smallest possible ledger number -
+// HistoryManager::GENESIS_LEDGER_SEQ
+uint32_t
+firstNeededLedger(CatchupConfiguration const& configuration)
+{
+    auto neededCount = std::max(configuration.count(), 1u);
+    return configuration.toLedger() > neededCount
+               ? configuration.toLedger() - neededCount + 1
+               : HistoryManager::GENESIS_LEDGER_SEQ;
+}
+
+std::pair<bool, uint32_t>
+computeCatchupStart(uint32_t smallestLedgerToApply,
+                    uint32_t smallestBucketApplyCheckpoint,
+                    HistoryManager const& historyManager)
+{
+    // checkpoint that contains smallestLedgerToApply - it is first one than
+    // can be applied, it is always greater than LCL
+    auto smallestCheckpointToApply =
+        checkpointContainingLedger(smallestLedgerToApply, historyManager);
+
+    // if first ledger that should be applied is on checkpoint boundary then
+    // we do an bucket-apply
+    if (smallestLedgerToApply == smallestCheckpointToApply)
+    {
+        return std::make_pair(true, smallestCheckpointToApply);
+    }
+
+    // we are before first checkpoint - applying buckets is not possible, so
+    // just apply transactions
+    if (smallestCheckpointToApply <= historyManager.getCheckpointFrequency())
+    {
+        return std::make_pair(false, smallestCheckpointToApply);
+    }
+
+    // we need to apply on previous checkpoint (if possible), so we are sure
+    // that we get required number of history entries in database
+    smallestCheckpointToApply -= historyManager.getCheckpointFrequency();
+    if (smallestCheckpointToApply >= smallestBucketApplyCheckpoint)
+    {
+        return std::make_pair(true, smallestCheckpointToApply);
+    }
+    else
+    {
+        // in that case we would apply before LCL, so there is no need to
+        // apply buckets
+        return std::make_pair(false, smallestBucketApplyCheckpoint);
+    }
+}
+}
+
+CatchupRange
+CatchupWork::makeCatchupRange(uint32_t lastClosedLedger,
+                              CatchupConfiguration const& configuration,
+                              HistoryManager const& historyManager)
+{
+    assert(lastClosedLedger > 0);
+    assert(configuration.toLedger() >= lastClosedLedger);
+    assert(configuration.toLedger() != CatchupConfiguration::CURRENT);
+
+    // smallest ledger that should be applied is either firstNeededLedger or
```

Comment:
rework this comment: "applied" is a bit vague, what you're computing is the maximum ledger number that we're willing to do "bucket apply" on, in order to replay enough of transaction history.